pkg/virt-controller/watch/pool.go

```diff
@@ -0,0 +1,988 @@
+	default:
+	}
+
+	return nil
+}
+
+func (c *PoolController) proactiveUpdate(pool *poolv1.VirtualMachinePool, vmUpdatedList []*virtv1.VirtualMachine) error {
+	var wg sync.WaitGroup
+	wg.Add(len(vmUpdatedList))
+	errChan := make(chan error, len(vmUpdatedList))
+	for i := 0; i < len(vmUpdatedList); i++ {
+		go func(idx int) {
+			defer wg.Done()
+			vm := vmUpdatedList[idx]
+
+			vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+			obj, exists, _ := c.vmiInformer.GetStore().GetByKey(vmiKey)
+			if !exists {
+				// no VMI to update
+				return
+			}
+			vmi := obj.(*virtv1.VirtualMachineInstance)
+			if vmi.DeletionTimestamp != nil {
+				// ignore VMIs which are already deleting
+				return
+			}
+
+			templateHash, templateOk := vm.Spec.Template.ObjectMeta.Labels[virtv1.VirtualMachineInstanceTemplateHash]
+			vmiHash, vmiOk := vmi.Labels[virtv1.VirtualMachineInstanceTemplateHash]
+			// compare the VMI's hash with the VM's hash
+			if templateOk && vmiOk && templateHash == vmiHash {
+				// nothing to do here, VMI is up-to-date with VM
+				return
+			}
+
+			// VMI and VM hash do not match, and we know the VM is up-to-date, so kill the VMI
+			err := c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Delete(vmi.ObjectMeta.Name, &v1.DeleteOptions{})
+			if err != nil {
+				c.recorder.Eventf(pool, k8score.EventTypeWarning, FailedUpdateVirtualMachineReason, "Error proactively updating VM %s/%s by deleting outdated VMI: %v", vm.Namespace, vm.Name, err)
+				errChan <- err
+			}
+			c.recorder.Eventf(pool, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Proactive update of VM %s/%s by deleting outdated VMI", vm.Namespace, vm.Name)
+		}(i)
+	}
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		// Only return the first error which occurred. We log the rest
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func (c *PoolController) update(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) syncError {
+	vmHash := hashVMTemplate(pool)
+	vmiHash := hashVMITemplate(pool)
+	// List of VMs that need to be updated
+	vmOutdatedList := []*virtv1.VirtualMachine{}
+	// List of VMs that are up-to-date that need to be checked to see if VMI is up-to-date
+	vmUpdatedList := []*virtv1.VirtualMachine{}
+
+	for _, vm := range vms {
+		outdated := false
+		if vm.Labels == nil {
+			outdated = true
+		} else {
+			oldHash, ok := vm.Labels[virtv1.VirtualMachineTemplateHash]
+			if !ok || oldHash != vmHash {
+				outdated = true
+			}
+		}
+
+		if outdated {
+			vmOutdatedList = append(vmOutdatedList, vm)
+		} else {
+			vmUpdatedList = append(vmUpdatedList, vm)
+		}
+	}
+
+	err := c.opportunisticUpdate(pool, vmOutdatedList, vmHash, vmiHash)
+	if err != nil {
+		return &syncErrorImpl{fmt.Errorf("Error during VM update: %v", err), FailedUpdateReason}
+	}
+
+	err = c.proactiveUpdate(pool, vmUpdatedList)
+	if err != nil {
+		return &syncErrorImpl{fmt.Errorf("Error during VMI update: %v", err), FailedUpdateReason}
+	}
```

Comment:
I'm probably missing something here, but from the design proposal it seemed to be as if `UpdateStrategy` can be either `Unmanaged`, `Opportunistic` or `Proactive`, but here it seems they are both tried sequentially.

Also, IIUC, the differences between them is that the opportunistic uses `c.clientset.VirtualMachine(vmCopy.Namespace).Update(vmCopy)` and proactive uses `c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Delete(vmi.ObjectMeta.Name, &v1.DeleteOptions{})`, I'm not sure I understand what's going on :)

I'm probably missing something, would appreciate elaboration

Reply:
Unmanaged means we never touch the VM or VMI
Opportunistic means the change is picked up on restart
Proactive means the change is picked up right away (causing a forced restart)

Because of the way our api is structured, the VM object represents the offline config, meaning it's the thing that stamps out a new VMI when a restart occurs. So an opportunistic update only gets applied to the VM (offline config), which then naturally causes an updated VMI (online config) after a restart. The proactive update is merely forcing the restart to occur after an opportunistic update.

Since the default behavior i'm implementing here is the `proactive` behavior, we perform the offline update (opportunistic) then force the online update (proactive) in sequence.

Reply:
Got it! Thanks for the explanation!